### ISSUE TITLE
Unhandled Exception: SqliteException(5): while executing statement, database is locked, database is locked (code 5)

### ISSUE THREAD

**Issue by isaanyoha** [simolus3/sqlite3.dart#157]
Hi @simolus3, thanks for this package!

Please kindly assist me here,

Why do I always get this issue 'Unhandled Exception: SqliteException(5): while executing statement, database is locked, database is locked (code 5)...'.   

This lock locks forever whenever it occurs. How do I fix this? My app requires concurrrency as it is a multiplayer game app. I am using:
- sqlite3: ^1.11.0
- testing on Android

![ezgif-3-774f9ss1e5sss5c2](https://user-images.githubusercontent.com/4920405/235490477-8f21de2a-b1cc-417e-8124-787e642741d7.png)

------DETAIL BELOW------
Initially, I had the issue on sqflite package and I thought it was about not using .dispose() so I rewrote the database side of the app to use sqlite3 and then use the prepare() & .dispose() functions for every single database call, yet it still happens. Please what should I do?

Again, long before now, the first version the app was written with native Android (Android Studio) and it used sqlite too and I did not have this issue on the code. But after I rewrote the app's new version on flutter, (using the same code pattern, alogrithm) and started to use the sqflite, and later sqlite3 the issue happens and the app database function stops and app game play stalls consequently.

I will be glad if I can get a solution to this.

Thank you!

**Comment by isaanyoha** [comment]
Also, I added
PRAGMA journal_mode=WAL;
PRAGMA busy_timeout=5000;

![ezgif-3-774fllll9ss1e5sss5c2](https://user-images.githubusercontent.com/4920405/235494691-c072fc95-d62b-43bd-b25f-e1c298fed6b4.png)

**Comment by simolus3** [comment]
Outside of that snippet, are you also still using the database in another location (e.g. in native code)?

**Comment by isaanyoha** [comment]
No, I don't think so. or maybe my app design pattern caused it to created it in another location?

Let me explain further of the database class...

The name of the class housing the database operation is DBController. Below is more snippet of the DBController class:

![111](https://user-images.githubusercontent.com/4920405/235544175-36ddd9ec-daa0-43ab-ae6a-9472459affdf.png)

![222](https://user-images.githubusercontent.com/4920405/235544164-aa8e05e6-653a-442f-a4f7-907d00896014.PNG)

![333](https://user-images.githubusercontent.com/4920405/235544171-1307e557-1dc1-47e2-90dc-b7d24352e850.PNG)

**Comment by isaanyoha** [comment]
Then this is the way I call the class...

First, I have a UI class and then I have a background service class which runs in an isolate
Below is one of the UI classes and how it uses the DBController class...

![666](https://user-images.githubusercontent.com/4920405/235545965-af62e5d5-77c0-4517-8e9a-938760d0e16e.png)

![555](https://user-images.githubusercontent.com/4920405/235545685-6119e7a1-0693-49b3-9fd8-5a2e8f8dd762.png)

**Comment by isaanyoha** [comment]
For the background service, I use a package called 'flutter_background_service: ^2.4.6'

**Comment by isaanyoha** [comment]
Heyyyy! I notice something o!

I added a counter 'databaseOpenCounter' to see if the database is called more than once on launching the app and it showed me that its been accessed more than one. How come then? Even with the singleton pattern?

![6661](https://user-images.githubusercontent.com/4920405/235551188-2e8ab5ec-f2ac-4ef8-9df9-e51afd5d7413.png)

![777](https://user-images.githubusercontent.com/4920405/235551199-0b0291bf-9ff2-43b4-b730-ba1f79c19f76.png)

**Comment by isaanyoha** [comment]
I notice that when the background service starts, opens another instance of the database. How do I go about this?

**Comment by isaanyoha** [comment]
I tried to use lock.synchronized, but it did not solve the issue as the background service is an isolate. 

So, to stop the database lock issue, can I have 2 instance of the database opened, where one is for the background service (isolate) and the other for other parts of the app but the both speaks to the same database? How do I do this or is there any other solution to resolve this?

**Comment by isaanyoha** [comment]
> Outside of that snippet, are you also still using the database in another location (e.g. in native code)?

So, I guess the answer to your question here is 'yes' as the background service which is also used in the app is an isolate (...and isolates are separate instance of the Dart Virtual Machine). 

What do you think is the way foward here?

**Comment by simolus3** [comment]
Using the database in two separate processes (or isolates for that matter) should not be a problem in general, and getting to conflicts should be pretty rare with `WAL`.

> I added a counter 'databaseOpenCounter' to see if the database is called more than once on launching the app and it showed me that its been accessed more than one. How come then? Even with the singleton pattern?

Well, there is a race condition in `DBController.database`: If it's called multiple times in a short time frame, it may open the database multiple times too (since all calls go past the condition). Excluding proper error handling, you could store the future and make sure it only gets invoked once:

```dart
Future<Database?>? _openingDatabase;

Future<Database?> get database => _openingDatabase ??= db();
```

But still, that doesn't explain the error.

**Comment by isaanyoha** [comment]
But how can I unlock the database? It hooks the app like forever. 

When this lock thing happens, it blocks other (insert, update, detele) sql operations (except select). and this lock locks forever without getting released.

How do I release myself when it occurs, atleast

![66dd61](https://user-images.githubusercontent.com/4920405/236113769-423c47aa-574e-4025-bdf5-692ee465a595.png)

Again, I restructed my app to now work without the background (so no isolate).

How can I release the locked database without restarting the app?

**Comment by simolus3** [comment]
> Ok, but how can I unlock the database?

sqlite3 will lock the database for writes only, the database will be unlocked afterwards. I think it's pretty weird that the database stays locked for so long, I've not seen this behavior before.

**Comment by isaanyoha** [comment]
Thanks for reply.

I use the sqlite3 package published on [https://pub.dev,](https://pub.dev/packages/sqlite3/)

With Flutter:
flutter pub add sqlite3

**Comment by simolus3** [comment]
`package:sqlite3` is a Dart package which supports Flutter, but also servers, CLI tools or any other non-Flutter based Dart application. To reliably access sqlite3 (the C library) on Flutter platforms though, it is best to include a copy with the application. This is where the `sqlite3_flutter_libs` package comes in. By depending on Flutter, it can include native build scripts to do just that.

So basically, you'll have to add a dependency on `sqlite3_flutter_libs` to get a recent sqlite3, but you don't have to change any of your Dart code.

**Comment by isaanyoha** [comment]
Ok, got that. Thank you!
